lib/schema.js

```diff
@@ -204,49 +341,107 @@ class Metaschema {
-    const keys = new Set([...defKeys, ...insKeys]);
-    for (const key of keys) {
-      const defKey = defKeys.has(key);
-      const insKey = insKeys.has(key);
-      if (!defKey && insKey) {
-        errors.push(`Field ${key} not defined`);
+    const schemaProps = new Set(Object.keys(schema));
+    const objectProps = new Set(Object.keys(object));
+    const props = new Set([...schemaProps, ...objectProps]);
+    for (const prop of props) {
+      const isSchemaProp = schemaProps.has(prop);
+      const isObjectProp = objectProps.has(prop);
+      if (isObjectProp && !isSchemaProp) {
+        errors.push(new ValidationError(
+          'unresolvedProperty', `${path}${prop}`
+        ));
         continue;
       }
-      const def = definition[key];
-      const val = instance[key];
-      const type = typeof val;
-      const domain = this.domains.get(def.domain);
-      if (!insKey && defKey) {
-        if (typeof def === 'function') {
-          const decorator = def.prototype.constructor.name;
-          if (decorator === 'Validate' && !def(instance)) {
-            errors.push('Validation failed');
-            continue;
-          }
+
+      const definition = schema[prop];
+
+      if (Object.getPrototypeOf(definition).name === 'Validate' && !patch) {
+        if (!definition(object)) {
+          errors.push(new ValidationError('validation', `${path}${prop}`));
         }
-        if (def.required) {
-          errors.push(`Field ${key} not found`);
+        continue;
+      }
+
+      if (!isObjectProp) {
+        if (definition.required && !patch) {
+          errors.push(new ValidationError(
+            'missingProperty', `${path}${prop}`
+          ));
         }
         continue;
       }
-      if (domain && type !== domain.type) {
-        errors.push(`Field ${key}:${val} expected to be ${domain.type}`);
+
+      if (definition.readOnly && patch) {
+        errors.push(new ValidationError(
+          'immutable', `${path}${prop}`
+        ));
+      }
+
+      const value = object[prop];
+
+      if (definition.truthy && !value) {
+        errors.push(new ValidationError(
+          'truthy', `${path}${prop}`
+        ));
       }
+
+      if (definition.domain) {
+        errors.push(
+          ...validateDomain(value, `${path}${prop}`, definition.definition)
+        );
+      } else if (definition.category) {
+        errors.push(
+          ...validateLink(value, `${path}${prop}`, definition, this, patch)
+        );
+      }
+    }
+    return errors;
+  }
+
+  // Validate instance against a category
+  //   categoryName - <string> category name
+  //   instance - <any> instance to validate
+  //   patch - <boolean>, flag to determine if the object contains patch or
+  //                     value, default: `false`
+  //   path - <string>, path to an object, for nested objects, default: `''`
+  // Returns: <MetaschemaError> | <null>
+  validate(categoryName, instance, patch = false, path = '') {
```

Comment:
Why do we have two methods `validate` and `[validate]`?
Maybe it's better to have different names?